smarts/env/wrappers/record_video.py

```diff
@@ -0,0 +1,166 @@
+#
+# pytype: disable=annotation-type-mismatch
+import os
+from typing import Callable
+
+import gym
+from gym import logger
+from gym.wrappers.monitoring import video_recorder
+
+
+def capped_cubic_video_schedule(episode_id):
+    """Util"""
+    if episode_id < 1000:
+        return int(round(episode_id ** (1.0 / 3))) ** 3 == episode_id
+    else:
+        return episode_id % 1000 == 0
+
+
+class RecordVideo(gym.Wrapper):
+    """A video recording wrapper."""
+    def __init__(
+        self,
+        env,
+        video_folder: str,
+        episode_trigger: Callable[[int], bool] = None,
+        step_trigger: Callable[[int], bool] = None,
+        video_length: int = 0,
+        name_prefix: str = "rl-video",
+    ):
+        super().__init__(env)
+
+        if episode_trigger is None and step_trigger is None:
+            episode_trigger = capped_cubic_video_schedule
+
+        trigger_count = sum(x is not None for x in [episode_trigger, step_trigger])
+        assert trigger_count == 1, "Must specify exactly one trigger"
+
+        self.episode_trigger = episode_trigger
+        self.step_trigger = step_trigger
+        self.video_recorder = None
+
+        self.video_folder = os.path.abspath(video_folder)
+        # Create output folder if needed
+        if os.path.isdir(self.video_folder):
+            logger.warn(
+                f"Overwriting existing videos at {self.video_folder} folder (try specifying a different `video_folder` for the `RecordVideo` wrapper if this is not desired)"
+            )
+        os.makedirs(self.video_folder, exist_ok=True)
+
+        self.name_prefix = name_prefix
+        self.step_id = 0
+        self.video_length = video_length
+
+        self.recording = False
+        self.recorded_frames = 0
+        self.is_vector_env = getattr(env, "is_vector_env", False)
+        self.episode_id = 0
+
+    def reset(self, **kwargs):
+        """Reset."""
+        observations = super().reset(**kwargs)
+        if not self.recording and self._video_enabled():
+            self.start_video_recorder()
+        return observations
+
+    def start_video_recorder(self):
+        """Start video."""
+        self.close_video_recorder()
+
+        video_name = f"{self.name_prefix}-step-{self.step_id}"
+        if self.episode_trigger:
+            video_name = f"{self.name_prefix}-episode-{self.episode_id}"
+
+        base_path = os.path.join(self.video_folder, video_name)
+        self.video_recorder = video_recorder.VideoRecorder(
+            env=self.env,
+            base_path=base_path,
+            metadata={"step_id": self.step_id, "episode_id": self.episode_id},
+        )
```

Comment:
`self.video_recorder` should definitely be encapsulated as `self._video_recorder`.